enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,561 @@
+
+* As a cluster administrator, having to keep track of a "boot" vs "live" image for a given cluster is not intuitive or user friendly. In the worst case scenario, I will have to reset a cluster(or do a lot of manual steps with rh-support in recovering the node) simply to be able to scale up nodes after an upgrade. If I'm managing a `MachineSet` managed cluster, once opted in, this feature will be a "switch on and forget" mechanism for me. If I'm managing a non `Machineset` managed cluster, this would provide me with documentation that I could follow after an upgrade to ensure my cluster has the latest bootimages.
+
+### Goals
+
+The MCO will take over management of the boot image references and the stub Ignition configuration. The installer is still responsible for creating the `MachineSet` at cluster bring-up, but once cluster installation is complete the MCO will ensure that boot images are in sync with the latest payload. From the user standpoint, this should cause less compatibility issues as nodes will no longer need to pivot to a different version of RHCOS during node scaleup.
+
+This should not interfere with existing workflows such as Hive and ArgoCD. As this is an opt-in mechanism, the cluster admin will be protected against such scenarios of accidental "reconciliation" and for additional safety, the MSBIC will also ensure that machinesets that have a valid OwnerReference will be excluded from boot image updates.
+
+### Non-Goals
+
+- The new subcontroller is only intended to support clusters that use MachineSet backed node scaling. This is meant to be a user opt-in feature, and if the user wishes to keep their boot images static it will let them do so.
+- This does not intend to solve [booting into custom pools](https://issues.redhat.com/browse/MCO-773). 
+- This does not target Hypershift, as [it does not use machinesets](https://github.com/openshift/hypershift/blob/32309b12ae6c5d4952357f4ad17519cf2424805a/hypershift-operator/controllers/nodepool/nodepool_controller.go#L2168).
+- This does not target [ControlPlaneMachineSets](https://docs.openshift.com/container-platform/4.14/machine_management/control_plane_machine_management/cpmso-about.html). This is considered future work and will be tracked by [MCO-1007](https://issues.redhat.com/browse/MCO-1007). 
+
+## Proposal
+
+__Overview__
+
+- The `machine-config-controller`(MCC) pod will gain a new sub-controller `machine_set_boot_image_controller`(MSBIC) that monitors `MachineSet` changes and the `coreos-bootimages` [ConfigMap](https://github.com/openshift/installer/pull/4760) changes.
+- Before processing a MachineSet, the MSBIC will check if the following conditions are satisfied:
+  - `ManagedBootImages` feature gate is active
+  - The cluster and/or the machineset is opted-in to boot image updates.
+  - The machineset does not have a valid owner reference. (eg. Hive, Cluster API and other managed machineset workflows)
```

Comment:
Thanks for the review @2uasimojo and the clarifaction of the `OwnerReference` 

The opt-in mechanism is simple if you want to opt-in the whole cluster(all machinesets). 
```
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
spec:
  managedBootImages:
    machineManagers:
    - resource: machinesets
      apiGroup: machine.openshift.io
      selection:
        mode: All
```
It does gets trickier if you want to enroll a subset of machinesets, which is when the label selector would come in. The above YAML snippet should work for most admins who would want these updates cluster wide and our goal is that it will be a one time action. Do you envision hive enrolling only a subset of machinesets?

I do have to caveat that shape of the API will be initially under TechPreview, so it subject to change. Once it goes GA - we will be committing to keep these options consistent through future releases.

We can definitely add an additional guard to check if the back door is enabled on the MSBIC side and exit with a log message. Question from my end, it looks like this is applied to MachinePools - I guess this isn't something that is visible on the MachineSet itself?

Reply:
> Do you envision hive enrolling only a subset of machinesets?

Ah, important point to clarify: I do not see *hive* managing this enrollment -- at least not directly. We like to minimize the number of in-cluster knobs we mirror explicitly on the hub, because it becomes a nightmare to try to keep up with differences across all the spoke OCP versions and infra providers we need to support. (Our management of MachineSets is itself a perfect example of this. Ask me about vsphere some time :eyeroll:) It also makes for a finicky and unintuitive support structure, where users must avoid managing those things directly on the spoke if they're using the hub mechanism... but not if they're not :P

Hive provides a [generic mechanism](https://github.com/openshift/hive/blob/master/docs/syncset.md) for users to create/sync/patch arbitrary objects from the hub to the spoke. I was expecting they would use that to opt into this feature. And as such, they would have the ability to shape it however they choose, including the most complex version of selector-ness you can think of. This is what I was thinking of when I said:

> I don't love the idea of hive trying to duplicate the logic MCO is using to process that thing.

...particularly since the thing, and the logic to process it, is likely to evolve over time across versions:

> I do have to caveat that shape of the API will be initially under TechPreview, so it subject to change. Once it goes GA - we will be committing to keep these options consistent through future releases.

====

> We can definitely add an additional guard to check if the back door is enabled on the MSBIC side and exit with a log message. Question from my end, it looks like this is applied to MachinePools - I guess this isn't something that is visible on the MachineSet itself?

Okay, let's play this out...

You are correct: Today the back door annotation is only on the MachinePool on hub side. But it wouldn't be tough to reflect that setting down into the MachineSets managed by that MachinePool.

As I mentioned earlier, I wouldn't expect MCO to pay attention to a `hive.openshift.io/*`-namespaced ("downstream") annotation/label... *but* if we designed an explicit opt-out annotation/label in the MCO namespace, hive could set _that_ thing on such MachineSets.

So theoretically you could have a MachinePool that manages MachineSet A with the "allow editing platform" knob; but `MachineConfiguration.spec.managedBootImages` is set up such that it would otherwise match MachineSets A, B, and C. In this case it would *not* twiddle A... and how would that manifest? A status condition somewhere? A log message?

By extension, could that same knob be used to explicitly opt *in*, where the MachineConfiguration would otherwise *not* select this MachineSet?

And, importantly, is that too weird/complex of a UX to design in? Because to be clear, we're talking about designing in a fully-supported knob on a MachineSet (or MachineDeployment in CAPI-land?) that overrides the behavior of the MachineConfiguration selector. Feels kind of icky. @JoelSpeed thoughts?

Reply:
Thanks for the detailed explanation and the link to `syncSet` mechanism! (: 

> So theoretically you could have a MachinePool that manages MachineSet A with the "allow editing platform" knob; but MachineConfiguration.spec.managedBootImages is set up such that it would otherwise match MachineSets A, B, and C. In this case it would not twiddle A... and how would that manifest? A status condition somewhere? A log message?

Our current error state is going to be a Prometheus metric based alert. So I suppose we could manifest this an unexpected opt-out annotation/label as an alert somehow.....but I'm not sure how popular those kind of alerts are in the Hive world. We were planning on adding a degrade on the MCO side once we got some mileage on the feature and see how customers interacted with it. At the moment, we don't think this feature warrants a degrade and blocking updates, so we went with an alert so it shows up on the console.

> And, importantly, is that too weird/complex of a UX to design in? Because to be clear, we're talking about designing in a fully-supported knob on a MachineSet (or MachineDeployment in CAPI-land?) that overrides the behavior of the MachineConfiguration selector. Feels kind of icky. @JoelSpeed thoughts?

Agreed...this isn't great. We'd essentially be mirroring our API efforts with a hive only annotation based approach. I don't love it, I'll let Joel weigh in. I'll have a think on this...

Reply:
Again, I don't think it's appropriate if it's a "hive only annotation". It would need to be a MCO-based, generic solution that hive just happens to be able to use to its advantage.

Re metrics/alerts: hive itself doesn't consume any such things from the spoke; but that's okay -- presumably these would be for the end user, not hive.

It occurred to me that we have this same consideration for the OwnerReference-based exclusion mechanism. What if MachineSet B has an OwnerReference? You would still need to surface that it has been excluded despite otherwise matching the configured selector.

Reply:
> Our current error state is going to be a Prometheus metric based alert

Huh? That's new to me, why prometheus and not something in cluster, it should be a condition on some object really.

> As I mentioned earlier, I wouldn't expect MCO to pay attention to a hive.openshift.io/*-namespaced ("downstream") annotation/label... but if we designed an explicit opt-out annotation/label in the MCO namespace, hive could set that thing on such MachineSets

The current API has a label selector within it. We could, in theory either enforce, or document that on Hive managed clusters that the `matchExpressions` section within the boot image updates is required to exclude machinesets with a well known hive label on them. That should be pretty straight forward from what I understand so far and won't need any API changes

Reply:
> Huh? That's new to me, why prometheus and not something in cluster, it should be a condition on some object really.

This was from the outcome of this thread: https://github.com/openshift/enhancements/pull/1496#discussion_r1489848889 
We're not 100% sure we want to degrade the operator just yet for an opt-in feature. We do want to eventually make it some sort of error state on the MCO side eventually though, based on admin usage

Reply:
I'm fine with an alert, but we should make sure there's a condition in the cluster as well

Reply:
Discussed briefly with David, just to clarify @JoelSpeed , are you asking for:
1. a general condition that something is wrong with the bootimage management in the MCO
2. a specific condition for machinesets that have management already (e.g. hive) that is being opted in

If it's the former, maybe the location that makes the most sense would be in the MachineConfiguration object as a condition for machineset management in general. And we can add that in as a requirement for GA via a follow-up API PR on top of the one we merged just now. WDYT?

Reply:
Yeah I think the condition should be alongside the configuration of the bootimages, on the same object, to show any errors there. It doesn't necessarily have to degrade the cluster (yet or ever), but, there should be some feedback on the object that the config is ok

Reply:
I can call this out in my next pass. The error state on the configuration object could also reflect an unexpected OwnerReference/Label on a machineset as well. 

Sorry about straying a bit off topic @2uasimojo! How do you feel about the approach suggested by @JoelSpeed  [here](https://github.com/openshift/enhancements/pull/1496#discussion_r1493070670)?

> > As I mentioned earlier, I wouldn't expect MCO to pay attention to a hive.openshift.io/*-namespaced ("downstream") annotation/label... but if we designed an explicit opt-out annotation/label in the MCO namespace, hive could set that thing on such MachineSets. 
>
> The current API has a label selector within it. We could, in theory either enforce, or document that on Hive managed clusters that the matchExpressions section within the boot image updates is required to exclude machinesets with a well known hive label on them. That should be pretty straight forward from what I understand so far and won't need any API changes

Reply:
IIUC you’re suggesting making the user responsible for excluding hive-managed MachineSets if they have the back door enabled? I’m okay with that in theory. Realistically nobody reads docs; but this back door isn’t documented anyway. This should be a small enough surface that this solution is acceptable.

Reply:
Chatted with @2uasimojo on slack a bit and we came up with some language to address the Hive concerns. I've plopped it [into a subsection](https://github.com/openshift/enhancements/blob/e4264f305e5ce65b5cfca365383bef387af317a6/enhancements/machine-config/manage-boot-images.md#variation-and-form-factor-considerations-optional) under Variation and form factor considerations, but I can move it if this isn't is the right place.